compiling on macOS

Hi @jguhlin,

Thanks for all your work on minimap2-rs!  I am working on a rust project that would be ideal for minimap2 bindings.  I have successfully integrated into my project in a linux environment, but compiling fails on a Mac.  I suspect that it is related to SSE4 code that I need to disable when successfully building minimap2 on a Mac using `make sse2only=1`.  I see you have related tasks in your todo list.  Have you gotten minimap2-rs to specifically compile on an ARM Mac?  Could you share the trick? :)  Or do you think something else is causing compilation failure?

Thanks for looking!!

I have output the compilation errors when running ` cargo build --release` below:


   Compiling minimap2-sys v0.1.10
   Compiling seq_io v0.3.1
   Compiling csv v1.2.1
   Compiling itertools v0.10.5
   Compiling toml v0.5.11
The following warnings were emitted during compilation:

warning: clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lpthread: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lpthread: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lpthread: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lpthread: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lpthread: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
warning: clang: warning: -lpthread: 'linker' input unused [-Wunused-command-line-argument]
warning: In file included from minimap2/ksw2_ll_sse.c:9:
warning: In file included from /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/emmintrin.h:13:
warning: In file included from /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/xmmintrin.h:13:
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:50:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_vec_init_v2si(__i, 0);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:129:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_packsswb((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:159:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_packssdw((__v2si)__m1, (__v2si)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:189:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_packuswb((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:216:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_punpckhbw((__v8qi)__m1, (__v8qi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:239:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_punpckhwd((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:260:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_punpckhdq((__v2si)__m1, (__v2si)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:287:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_punpcklbw((__v8qi)__m1, (__v8qi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:310:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_punpcklwd((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:331:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_punpckldq((__v2si)__m1, (__v2si)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:352:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_paddb((__v8qi)__m1, (__v8qi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:373:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_paddw((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:394:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_paddd((__v2si)__m1, (__v2si)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:416:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_paddsb((__v8qi)__m1, (__v8qi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:439:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_paddsw((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:461:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_paddusb((__v8qi)__m1, (__v8qi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:483:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_paddusw((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:504:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_psubb((__v8qi)__m1, (__v8qi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:525:12: error: invalid conversion between vector type '__m64' (vector of 1 'long long' value) and integer type 'int' of different size
warning:     return (__m64)__builtin_ia32_psubw((__v4hi)__m1, (__v4hi)__m2);
warning:            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: fatal error: too many errors emitted, stopping now [-ferror-limit=]
warning: 20 errors generated.

error: failed to run custom build command for `minimap2-sys v0.1.10`

> Hey thanks. Could you post your error logs?

The error messages were exactly the same as encountered by @scfurl. Without simde the one at the top, and with simde the one further below (Undefined symbols for architecture arm64...).

I also tried loading the crate directly via GitHub. When doing so I have to manually turn default-features off and separately specify flate2 (in the chopper .toml) in addition to simde (as a feature of minimap2), only to end up at the latter error message (Undefined symbols...).

Cool thanks. Good to know. I'll dig into it more next week. Cheers

On Tue, 18 Jul 2023, 4:14 pm leiste375, ***@***.***> wrote:

> Hey thanks. Could you post your error logs?
>
> The error messages were exactly the same as encountered by @scfurl
> <https://github.com/scfurl>. Without simde the one at the top, and with
> simde the one further below (Undefined symbols for architecture arm64...).
>
> I also tried loading the crate directly via GitHub. When doing so I have
> to manually turn default-features off and separately specify flate2 (in the
> chopper .toml) in addition to simde (as a feature of minimap2), only to end
> up at the latter error message (Undefined symbols...).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jguhlin/minimap2-rs/issues/30#issuecomment-1639555523>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOA2CGRCQ22THDTNZBLIGDXQYSSVANCNFSM6AAAAAAWQOEIJY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

--- a/minimap2-sys/build.rs
+++ b/minimap2-sys/build.rs
@@ -110,6 +110,7 @@ fn configure(mut cc: &mut cc::Build) {
     target_specific(&mut cc);
 }
 
+/*
 #[cfg(all(target_arch = "aarch64", feature = "neon"))]
 fn target_specific(cc: &mut cc::Build) {
     cc.include("minimap2/sse2neon/");
@@ -126,9 +127,12 @@ fn target_specific(cc: &mut cc::Build) {
     cc.flag("-D_FILE_OFFSET_BITS=64");
     cc.flag("-fsigned-char");
 }
+*/
 
-#[cfg(all(target_arch = "aarch64", not(feature = "neon")))]
+//Include neon for all "aarch64".
+#[cfg(all(target_arch = "aarch64"))]
 fn target_specific(cc: &mut cc::Build) {
+    cc.include("minimap2/sse2neon/");
     // For aarch64 targets with neon
     // Add the following files:
     // ksw2_extz2_neon.o ksw2_extd2_neon.o ksw2_exts2_neon.o
@@ -137,10 +141,12 @@ fn target_specific(cc: &mut cc::Build) {
     cc.file("minimap2/ksw2_exts2_sse.c");
     cc.flag("-DKSW_SSE2_ONLY");
 
-    // CFLAGS+=-D_FILE_OFFSET_BITS=64 -mfpu=neon -fsigned-char
+    // CFLAGS+=-D_FILE_OFFSET_BITS=64 -fsigned-char -Isse2neon -D__SSE2__
     cc.flag("-D_FILE_OFFSET_BITS=64");
-    cc.flag("-mfpu=neon");
     cc.flag("-fsigned-char");
+    cc.flag("-Isse2neon");
+    cc.flag("-D__SSE2__");
+
 }
 
 #[cfg(target_arch = "x86_64")]

